synapse/storage/databases/main/receipts.py

```diff
@@ -133,38 +133,63 @@ async def get_receipts_for_room(
             desc="get_receipts_for_room",
         )
 
-    @cached(num_args=3)
```

Comment:
I think we need to keep `get_last_receipt_event_id_for_user` and `get_receipts_for_user` as cached methods, receipts get fetched quite often.

Reply:
I'm suspecting you got rid of these because you want to be able to invalidate a cache using on partial arguments -- the way to do this is to setup the cache as a tree:

```python
@cached(tree=True)
async def get_last_receipt_event_id_for_user(self, user_id, room_id, receipt_type):
    ...
```

Then when you invalidate it you can give partial arguments:

```python
get_last_receipt_event_id_for_user.invalidate((user_id, room_id,))
```

Reply:
That doesn't seem to work as `List` isn't hashable :/

Reply:
Right, most things aren't. I guess I see two options:

1. Use something hashable (a `tuple` is easier, but a `frozenset` is likely more correct).
2. Modify the methods to be read receipts.

I suspect going the second route is easier, but I would be OK with either.

Reply:
It seems that doesn't work? When `invalidate()` is called, we would need to invalidate the list of RR types but also all other lists of RR types which include some of the same elements? That does not seem to happen - am I missing something?